File: python/data_structures/heap/max_heap.py

class MaxHeap:
    def __init__(self):
        self._data = [] 
        
    def _parent(self, i):
        return (i - 1) // 2
    
    def _left(self, i):
        return i * 2 + 1
    
    def _right(self, i):
        return i * 2 + 2
    
    def insert(self,value):
        self._data.append(value)
        self._heapify_up(len(self._data) - 1)
    
    def pop(self):
        if not self._data:
            raise IndexError('Heap empty')

        if len(self._data) == 1:
            return self._data.pop()
        
        root = self._data[0]

        self._data[0] = self._data.pop()
        self._heapify_down(0)
        return root


    def _heapify_up(self, i):
        while i > 0 and self._data[i] > self._data[self._parent(i)]:
            p = self._parent(i)
            self._data[i], self._data[p] = self._data[p], self._data[i]
            i = p
        
    def _heapify_down(self, i):
        biggest = i
        left = self._left(i)
        right = self._right(i)

        if left < len(self._data) and self._data[left] > self._data[biggest]:
            biggest = left
        if right < len(self._data) and self._data[right] > self._data[biggest]:
            biggest = right
        
        if biggest != i: 
            self._data[i], self._data[biggest] = self._data[biggest], self._data[i]
            self._heapify_down(biggest)



           

File: java/data_structures/heap/MaxHeap.java
package data_structures.heap;

import java.util.*; 

public class MaxHeap {

	private final List<Integer> data = new ArrayList<>();
	
	
	// --- Public API ---
    public void insert(int value) {
        data.add(value);
        heapifyUp(data.size() - 1);
    }

    public int pop() {
        if (data.isEmpty()) throw new NoSuchElementException("Heap is empty");
        int root = data.get(0);
        int lastIdx = data.size() - 1;

        // Move last to root, remove tail, then restore heap property.
        swap(0, lastIdx);
        data.remove(lastIdx);
        if (!data.isEmpty()) heapifyDown(0);

        return root;
    }

    public int peek() {
        if (data.isEmpty()) throw new NoSuchElementException("Heap is empty");
        return data.get(0);
    }

    public int size() {
        return data.size();
    }
	
	// --- Internals ---
    private static int parent(int i) { return (i - 1) / 2; }
    private static int left(int i)   { return 2 * i + 1; }
    private static int right(int i)  { return 2 * i + 2; }

    private void heapifyUp(int i) {
        // Bubble up while current > parent.
        while (i > 0) {
            int p = parent(i);
            if (data.get(i) <= data.get(p)) break;
            swap(i, p);
            i = p;
        }
    }

    private void heapifyDown(int i) {
        // Sift down while there is at least a left child.
        int n = data.size();
        while (left(i) < n) {
            int l = left(i), r = right(i);
            int largest = i;

            if (data.get(l) > data.get(largest)) largest = l;
            if (r < n && data.get(r) > data.get(largest)) largest = r;

            if (largest == i) break;
            swap(i, largest);
            i = largest;
        }
    }

    private void swap(int i, int j) {
        int tmp = data.get(i);
        data.set(i, data.get(j));
        data.set(j, tmp);
    }
}
